### PR TITLE
Accept Invitations to Organization

### PR DESCRIPTION
**Pull request by Colin-Lane**
@ryanpbrewster @leogrim @yingjieMiao

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala [hunk 1]
```diff
@@ -212,8 +212,37 @@ class OrganizationInviteCommanderImpl @Inject() (db: Database,
...
+    (membershipOpt match {
+      case Some(membership) =>
+        val modifyRequests = sortedInvitations.map(currentInvitation => OrganizationMembershipModifyRequest(orgId, currentInvitation.inviterId, userId, currentInvitation.role))
+        val findFirstSuccess = modifyRequests.toStream.map(organizationMembershipCommander.modifyMembership(_))
+          .find(_.isRight) // stop at first success
+        findFirstSuccess.map(_.right.map(_.membership))
+          .getOrElse(Left(OrganizationFail.NO_VALID_INVITATIONS))
+      case None => // new membership
+        val addRequests = sortedInvitations.map(currentInvitation => OrganizationMembershipAddRequest(orgId, currentInvitation.inviterId, userId, currentInvitation.role))
+        val findFirstSuccess = addRequests.toStream.map(organizationMembershipCommander.addMembership(_))
```
Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: I need a method that can modify a membership and pick the best valid invitation instead of this.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala [hunk 2]
```diff
@@ -212,8 +212,37 @@ class OrganizationInviteCommanderImpl @Inject() (db: Database,
...
+          .find(_.isRight)
+        findFirstSuccess.map(_.right.map(_.membership))
+          .getOrElse(Left(OrganizationFail.NO_VALID_INVITATIONS))
+    }).right.map { success =>
+      // on success accept invitations
+      db.readWrite { implicit s =>
+        // TODO: notify inviters about invitee accepting invitation.
+        invitations.map(_.copy(state = OrganizationInviteStates.ACCEPTED)).foreach(organizationInviteRepo.save(_))
+      }
```
Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:251]: TODO: On successful acceptance of invitation notify all distinct inviters about the new user.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala (hunk 1 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: What if we change `OrganizationMembershipAddRequest` (and `ModifyRequest`) to take in `roles: OrganizationRole*` and it sorts them by role and finds the first one that succeeds.

I guess then it'd give back the `OrganizationMembershipAddResponse` with just the request that actually passed.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala [hunk 3]
```diff
@@ -212,8 +212,37 @@ class OrganizationInviteCommanderImpl @Inject() (db: Database,
...
+    }
+
+    val sortedInvitations = invitations.sortBy(_.role).reverse
+    (membershipOpt match {
+      case Some(membership) =>
+        val modifyRequests = sortedInvitations.map(currentInvitation => OrganizationMembershipModifyRequest(orgId, currentInvitation.inviterId, userId, currentInvitation.role))
+        val findFirstSuccess = modifyRequests.toStream.map(organizationMembershipCommander.modifyMembership(_))
+          .find(_.isRight) // stop at first success
+        findFirstSuccess.map(_.right.map(_.membership))
+          .getOrElse(Left(OrganizationFail.NO_VALID_INVITATIONS))
```
Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: Do we want to handle promoting existing members here, or just do it automatically on invitation send?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala (hunk 1 above)
Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: That may not work. ex.
An `Admin` invites you to `Admin` with him.
A `Member` invites you to `Member` with him.
You accept the `Member` invitation.
If `ModifyRequest` takes in the member inviter and a `Seq[Role]` then it may fail thinking the `Member` tried to invite someone to `Admin`.
In addition to the `Role` it would also need to take in the `inviterId` as a `Seq[(Role, Inviter)]`.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala (hunk 3 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: I think we just do it on invitation send. That way it's basically a convenience for organization admins, but otherwise it behaves exactly the same as normal promotions.

Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: Then I will remove this logical branch.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala [hunk 4]
```diff
@@ -212,8 +212,37 @@ class OrganizationInviteCommanderImpl @Inject() (db: Database,
...
+      val allInvitations = authInvitations ++ userInvitations
+      val existingMembership = organizationMembershipRepo.getByOrgIdAndUserId(orgId, userId)
+      (allInvitations, existingMembership)
+    }
+
+    val sortedInvitations = invitations.sortBy(_.role).reverse
+    (membershipOpt match {
+      case Some(membership) =>
+        val modifyRequests = sortedInvitations.map(currentInvitation => OrganizationMembershipModifyRequest(orgId, currentInvitation.inviterId, userId, currentInvitation.role))
+        val findFirstSuccess = modifyRequests.toStream.map(organizationMembershipCommander.modifyMembership(_))
```
Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: going to rename this to `firstSuccess` instead of `findFirstSuccess`

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: I don't think that's good coupling. You add someone with a defined role, not multiple. The issue isn't with all the roles here, it seems like it's with the fact that an inviter's rights may change and a valid invitation may no longer lead to a valid `OrganizationMembershipAddRequest`. I think the simple way to handle this is to check the inviter's rights when the invite is sent and forget about it afterwards. Then when an authentic (as in valid `authToken`) invite is accepted, grab the max role from all other invites and create the membership (don't check inviters' rights again). Either go straight to the repo or make `OrganizationMembershipAddRequest.requesterId` optional (I would go with that last one).

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: So the other options are: we basically move all of this logic into its own method in `MembershipCommander`, or we write a `addMultipleMemberships(requests: Seq[OrganizationMembershipAddRequest]): Seq[OrganizationMembershipAddResponse]`. My concern about the latter is what the behavior would be if you passed in a bunch of requests from different requesters, for different targets, some of which fail and some of which succeed. How do you even return that? Is it worth the complexity?

Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: The current implementation will only check the next invitation if the current does not succeed (we are using `toStream` and `find`). This makes the additional overhead minor for 99% of invitations (how often does someone invite and then get demoted?).
The alternative of using an `optional requestorId` would eliminate any overhead, but then we have introduced an extra `Option` into our code which can lead to dangerous behavior when changing someone's role. I think it is a good idea to keep the requestor id required when modifying someone's role.

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: Then you need to handle the case of an invite being no longer valid, and let the invitee know when this happens - we probably need this anyway for invites that are explicitly revoked. Additionally, note that after accepting an invite for a given role, users may end up with a lesser one.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala [hunk 5]
```diff
@@ -212,8 +218,68 @@ class OrganizationInviteCommanderImpl @Inject() (db: Database,
...
+          .getOrElse(Left(OrganizationFail.NO_VALID_INVITATIONS))
+    }).right.map { success =>
+      // on success accept invitations
+      db.readWrite { implicit s =>
+        // Notify inviters on organization joined.
+        notifyInviterOnOrganizationInvitationAcceptance(invitations, userRepo.get(userId), organizationRepo.get(orgId))
+        invitations.map(_.copy(state = OrganizationInviteStates.ACCEPTED)).foreach(organizationInviteRepo.save(_))
+      }
+      success
+    }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:253]: Can we use use an intermediate `val` to split this block of logic into smaller pieces? Right now the entire flow is a little hard to follow.

Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:253]: Yeah, I guess this is what you would call a run-on-sentence for programmers. I can split it.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationMembershipCommander.scala [hunk 6]
```diff
@@ -136,8 +136,8 @@ class OrganizationMembershipCommanderImpl @Inject() (
       if (validRequest(request)) {
         val membership = organizationMembershipRepo.getByOrgIdAndUserId(request.orgId, request.targetId).get
         val org = organizationRepo.get(request.orgId)
-        organizationMembershipRepo.save(org.modifiedMembership(membership, request.newRole))
-        Right(OrganizationMembershipModifyResponse(request))
+        val newMembership = organizationMembershipRepo.save(org.modifiedMembership(membership, request.newRole))
+        Right(OrganizationMembershipModifyResponse(request, newMembership))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationMembershipCommander.scala:140]: I like these. Maybe we should follow the same pattern in `removeMembership`.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala (hunk 5 above)
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:253]: is this just `invitations.foreach{  organizationInviteRepo.save(_.copy(state = OrganizationInviteStates.ACCEPTED)) }` ?

Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:253]: Kind of, but to make it compile it would need to be `invitations.foreach{ invite => organizationInviteRepo.save(invite.copy(state = OrganizationInviteStates.ACCEPTED)) }`

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala [hunk 7]
```diff
@@ -212,8 +218,68 @@ class OrganizationInviteCommanderImpl @Inject() (db: Database,
...
+        linkUrl = s"https://www.kifi.com/${invitee.username.value}",
+        imageUrl = inviteeImage,
+        sticky = false,
+        category = NotificationCategory.User.ORGANIZATION_JOINED,
+        extra = Some(Json.obj(
+          "member" -> BasicUser.fromUser(invitee),
+          "organization" -> Json.toJson(OrganizationNotificationInfo.fromOrganization(org, orgImageOpt))
+        ))
+      ) map { _ =>
+          val canSendPush = kifiInstallationCommander.isMobileVersionEqualOrGreaterThen(inviterId, KifiAndroidVersion("2.2.4"), KifiIPhoneVersion("2.1.0"))
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: does this second push notification HAS to happen after the global notification? Can we fire two of them at the same time? Any dependency I am missing?

Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: That is an interesting question. `sendGlobalNotification` is a `Future`, it may have something to do with exceptions being thrown in the first one (being caught as `SafeFuture`), then not sending to the second. I will have to ask why `LibraryInviteCommander` did it this way.

Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: does global notification mean the notification on the extension? At its current state, it just enforces no mobile notification can be pushed until extension gets notified. But it doesn't enforce mobile is sent as well.

Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: I am not sure. @andrewconner @leogrim  what is the reason behind why `LibraryInviteCommander:100` maps on the future rather than sending both at once, and should we maybe do it differently for `OrganizationInviteCommander`?

Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: Actually not sure. I was involved in a refactor recently but didn't write it. Should be fine to do both at once.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala (hunk 5 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:253]: Indeed, you can use a single `foreach` with your syntax above is correct.
As a rule of thumb, whenever you use `foreach` if should feel fishy to use it with `map` or `flatMap`, because you're side-effecting anyway.
Similarly, if you use `map` and `filter` together, you can often use `collect` nicely instead.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala [hunk 8]
```diff
@@ -212,8 +218,69 @@ class OrganizationInviteCommanderImpl @Inject() (db: Database,
...
+      val userInvitations = organizationInviteRepo.getByOrgAndUserId(orgId, userId)
+      val existingMembership = organizationMembershipRepo.getByOrgIdAndUserId(orgId, userId)
+      (userInvitations, existingMembership)
+    }
+
+    val sortedInvitations = invitations.sortBy(_.role).reverse
+    val updatedMembership: Either[OrganizationFail, OrganizationMembership] = membershipOpt match {
+      case Some(membership) => Right(membership) // already a member
+      case None => // new membership
+        val addRequests = sortedInvitations.map(currentInvitation => OrganizationMembershipAddRequest(orgId, currentInvitation.inviterId, userId, currentInvitation.role))
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala]: You could inline `sortedInvitations` here, or at least move its declaration in here - it doesn't need to be in the outside scope, away from where it's actually used and clouding the intent of the main flow. (I agree this is small, but it makes for good habits to hide intermediate variables inside inner scopes)

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationView.scala [hunk 9]
```diff
@@ -1,9 +1,31 @@
...
 import com.keepit.common.db.Id
 import com.keepit.common.mail.EmailAddress
+import com.keepit.common.store.ImagePath
+import com.kifi.macros.json
 import play.api.http.Status._
 
+@json case class OrganizationImageInfo(path: ImagePath, x: Int, y: Int)
+
+object OrganizationImageInfo {
+  def createInfo(img: OrganizationAvatar) = OrganizationImageInfo(img.imagePath, img.position.map(_.x).getOrElse(50), img.position.map(_.y).getOrElse(50))
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationView.scala]: `OrganizationAvatar` doesn't need a position, my mistake when I designed the initial model. I think I updated the Team Up document since then.

File: kifi-backend/modules/shoebox/test/com/keepit/commanders/OrganizationInviteCommanderTest.scala [hunk 10]
```diff
@@ -119,5 +119,86 @@ class OrganizationInviteCommanderTest extends TestKitSupport with SpecificationL
...
+            UserFactory.user().withId(userId).saved
+            val org = inject[OrganizationRepo].save(Organization(name = "kifi", ownerId = inviterId, handle = None))
+            memberRepo.save(org.newMembership(userId = inviterId, role = OrganizationRole.MEMBER))
+            inviteRepo.save(OrganizationInvite(organizationId = org.id.get, inviterId = inviterId, userId = Some(userId), role = OrganizationRole.OWNER))
+            org
+          }
+          inviteCommander.acceptInvitation(org.id.get, userId) === Left(OrganizationFail.NO_VALID_INVITATIONS)
+        }
+      }
+    }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/test/com/keepit/commanders/OrganizationInviteCommanderTest.scala:202]: Nice testing :+1:

**Comment by leogrim** [comment]
A few comments to look at, good job overall!

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala (hunk 5 above)
Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:253]: Excellent point about the end result side-effecting, that makes sense.
